packages/hydrogen/src/analytics-manager/AnalyticsProvider.tsx

```diff
@@ -0,0 +1,524 @@
+  prevCart: Awaited<AnalyticsProviderProps['cart']>;
+};
+
+export type AnalyticsContextValue = {
+  /** A function to tell you the current state of if the user can be tracked by analytics. Defaults to Customer Privacy API's `window.Shopify.customerPrivacy.userCanBeTracked()`. */
+  canTrack: NonNullable<AnalyticsProviderProps['canTrack']>;
+  /** The current cart state. */
+  cart: Awaited<AnalyticsProviderProps['cart']>;
+  /** The custom data passed in from the `AnalyticsProvider`. */
+  customData?: AnalyticsProviderProps['customData'];
+  /** The previous cart state. */
+  prevCart: Awaited<AnalyticsProviderProps['cart']>;
+  /** A function to publish an analytics event. */
+  publish: typeof publish;
+  /** A function to register with the analytics provider. */
+  register: (key: string) => {ready: () => void};
+  /** The shop configuration required to publish events to Shopify. */
+  shop: Awaited<AnalyticsProviderProps['shop']>;
+  /** A function to subscribe to analytics events. */
+  subscribe: typeof subscribe;
+};
+
+export const defaultAnalyticsContext: AnalyticsContextValue = {
+  canTrack: () => false,
+  cart: null,
+  customData: {},
+  prevCart: null,
+  publish: () => {},
+  shop: null,
+  subscribe: () => {},
+  register: () => ({ready: () => {}}),
+};
+
+const AnalyticsContext = createContext<AnalyticsContextValue>(
+  defaultAnalyticsContext,
+);
+
+const subscribers = new Map<
+  string,
+  Map<string, (payload: EventPayloads) => void>
+>();
+const registers: Record<string, boolean> = {};
+
+function areRegistersReady() {
+  let ready = true;
+  Object.keys(registers).forEach((key) => {
+    if (!registers[key]) {
+      ready = false;
+    }
+  });
+  return ready;
+}
+
+// Overload functions for each subscribe event
+function subscribe(
+  event: typeof AnalyticsEvent.PAGE_VIEWED,
+  callback: (payload: PageViewPayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.PRODUCT_VIEWED,
+  callback: (payload: ProductViewPayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.COLLECTION_VIEWED,
+  callback: (payload: CollectionViewPayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.CART_VIEWED,
+  callback: (payload: CartViewPayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.SEARCH_VIEWED,
+  callback: (payload: SearchViewPayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.CART_UPDATED,
+  callback: (payload: CartUpdatePayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.PRODUCT_ADD_TO_CART,
+  callback: (payload: CartLineUpdatePayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.PRODUCT_REMOVED_FROM_CART,
+  callback: (payload: CartLineUpdatePayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.CUSTOM_EVENT,
+  callback: (payload: CustomEventPayload) => void,
+): void;
```

Comment:
Curious, can we simplify this with something like the following? Or is there any downside?

<img width="726" alt="image" src="https://github.com/Shopify/hydrogen/assets/1634092/74b506eb-8b25-4427-a670-e6044f945577">

Reply:
The same goes for the `publish` function types.

Reply:
Go stuck on 

```
type EventPayload = {
  page_viewed: PageViewPayload;
  product_viewed: ProductViewPayload;
  collection_viewed: CollectionViewPayload;
  cart_viewed: CartViewPayload;
  search_viewed: SearchViewPayload;
  cart_updated: CartUpdatePayload;
  product_added_to_cart: CartLineUpdatePayload;
  product_removed_from_cart: CartLineUpdatePayload;
  [`custom_${string}`]: CustomEventPayload;   // variable key that typescript doesn't like
};
```